go/nbs/conjoiner.go

```diff
@@ -0,0 +1,201 @@
+	// currently needed; callers may be working with an out-of-date notion of
+	// upstream state. |novelCount|, the number of new tables the caller is
+	// trying to land, may be used in this determination. Implementations must
+	// be goroutine-safe.
+	Conjoin(mm manifest, p tablePersister, novelCount int, stats *Stats)
+}
+
+func newAsyncConjoiner(maxTables int) *asyncConjoiner {
+	return &asyncConjoiner{
+		waiters:   map[string]chan struct{}{},
+		maxTables: maxTables,
+	}
+}
+
+type asyncConjoiner struct {
+	mu        sync.RWMutex
+	waiters   map[string]chan struct{}
+	maxTables int
+}
+
+func (c *asyncConjoiner) ConjoinRequired(ts tableSet) bool {
+	return ts.Size() > c.maxTables
+}
+
+// Conjoin checks to see if there's already a conjoin underway for the store
+// described by |mm|. If so, it blocks until that conjoin completes. If not,
+// it starts one and blocks until it completes. Conjoin can be called
+// concurrently from many goroutines.
+func (c *asyncConjoiner) Conjoin(mm manifest, p tablePersister, novelCount int, stats *Stats) {
+	needsConjoin := func(upstreamCount int) bool {
+		return upstreamCount+novelCount > c.maxTables
+	}
+	c.await(mm.Name(), func() { conjoin(mm, p, needsConjoin, stats) }, nil)
+	return
+}
+
+// await checks to see if there's already something running for |id| and, if
+// so, waits for it to complete. If not, it runs f() and waits for it to
+// complete. While f() is running, other callers to await that pass in |id|
+// will block until f() completes.
+func (c *asyncConjoiner) await(id string, f func(), testWg *sync.WaitGroup) {
+	wait := func() <-chan struct{} {
+		c.mu.Lock()
+		defer c.mu.Unlock()
+
+		if ch, present := c.waiters[id]; present {
+			return ch
+		}
+		c.waiters[id] = make(chan struct{})
+		go c.runAndNotify(id, f)
+		return c.waiters[id]
+	}()
+	if testWg != nil {
+		testWg.Done()
+	}
+	<-wait
+}
+
+// runAndNotify runs f() and, upon completion, signals everyone who called
+// await(id).
+func (c *asyncConjoiner) runAndNotify(id string, f func()) {
+	f()
+
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	ch, present := c.waiters[id]
+	d.PanicIfFalse(present)
+	close(ch)
+	delete(c.waiters, id)
+}
+
+func conjoin(mm manifest, p tablePersister, needsConjoin func(int) bool, stats *Stats) {
+	b := &backoff.Backoff{
+		Min:    128 * time.Microsecond,
+		Max:    10 * time.Second,
+		Factor: 2,
+		Jitter: true,
+	}
+
+	exists, _, lock, root, upstream := mm.ParseIfExists(nil)
+	d.PanicIfFalse(exists)
+	// This conjoin may have been requested by someone with an out-of-date notion of what's upstream. Verify that we actually still believe a conjoin is needed and, if not, return early
+	if !needsConjoin(len(upstream)) {
+		return
+	}
+
+	var conjoined tableSpec
+	var conjoinees, keepers []tableSpec
+
+	for {
+		if conjoinees == nil {
+			conjoined, conjoinees, keepers = conjoinTables(p, upstream, stats)
+		}
+
+		specs := append(make([]tableSpec, 0, len(keepers)+1), conjoined)
```

Comment:
I wish :-/ It complains that there are too many arguments to append if you try that.